src/proxy-restore.c

```diff
@@ -11,7 +11,7 @@ static SEXP fns_vec_restore_dispatch = NULL;
 SEXP vec_restore_default(SEXP x, SEXP to) {
   SEXP attrib = ATTRIB(to);
 
-  if (attrib == R_NilValue) {
+  if (attrib == R_NilValue && !IS_S4_OBJECT(to)) {
```

Comment:
If we support S4 objects here, it should be to throw an internal error recommending the class author to implement a restore method.

Reply:
Very minor, but you could do `bool is_s4 = IS_S4_OBJECT(to)` since you call `IS_S4_OBJECT(to)` later on as well